subsys/dfu/dfu_target/src/dfu_target_mcuboot.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/*
+ * Ensure 'strnlen' is available even with -std=c99. If
+ * _POSIX_C_SOURCE was defined we will get a warning when referencing
+ * 'strnlen'. If this proves to cause trouble we could easily
+ * re-implement strnlen instead, perhaps with a different name, as it
+ * is such a simple function.
+ */
+#if !defined(_POSIX_C_SOURCE)
+#define _POSIX_C_SOURCE 200809L
+#endif
+#include <string.h>
+
+#include <zephyr.h>
+#include <pm_config.h>
+#include <logging/log.h>
+#include <dfu/mcuboot.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_stream.h>
+
+LOG_MODULE_REGISTER(dfu_target_mcuboot, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+#define MAX_FILE_SEARCH_LEN 500
+#define MCUBOOT_HEADER_MAGIC 0x96f3b83d
+#define MCUBOOT_SECONDARY_LAST_PAGE_ADDR \
+	(PM_MCUBOOT_SECONDARY_ADDRESS + PM_MCUBOOT_SECONDARY_SIZE - 1)
+
+
+static uint8_t *stream_buf;
+static size_t stream_buf_len;
+
+int dfu_ctx_mcuboot_set_b1_file(const char *file, bool s0_active,
+				const char **update)
+{
+	if (file == NULL || update == NULL) {
+		return -EINVAL;
+	}
+
+	/* Ensure that 'file' is null-terminated. */
+	if (strnlen(file, MAX_FILE_SEARCH_LEN) == MAX_FILE_SEARCH_LEN) {
+		return -ENOTSUP;
+	}
+
+	/* We have verified that there is a null-terminator, so this is safe */
+	char *space = strstr(file, " ");
+
+	if (space == NULL) {
+		/* Could not find space separator in input */
+		*update = NULL;
+
+		return 0;
+	}
+
+	if (s0_active) {
+		/* Point after space to 'activate' second file path (S1) */
+		*update = space + 1;
+	} else {
+		*update = file;
+
+		/* Insert null-terminator to 'activate' first file path (S0) */
+		*space = '\0';
+	}
+
+	return 0;
+}
+
+bool dfu_target_mcuboot_identify(const void *const buf)
+{
+	/* MCUBoot headers starts with 4 byte magic word */
+	return *((const uint32_t *)buf) == MCUBOOT_HEADER_MAGIC;
+}
+
+int dfu_target_mcuboot_set_buf(uint8_t *buf, size_t len)
```

Comment:
Why do we want the `set_buf` function. It seems to fit more naturally into `init`. If it is due to not adding more parameters to `init`  I think we should just add it and allow it to be set to `NULL`  in use-cases where it is not needed.

Reply:
> If it is due to not adding more parameters to init

Perhaps this would align nicely with my [suggestion of an init struct](https://github.com/nrfconnect/sdk-nrf/pull/3422/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h&hide-deleted-files=true#r533359236)?

Reply:
So the thing is that the init function is automatically invoked by the dfu_target library, and that requires the API of all the _init functions of all the dfu target implementations to be aligned. 

Because of that we have some additional initialization functions, such as this one.

Reply:
Where is this used? Could it instead be a static buffer with size set by kconfig?

Reply:
I prefer that the user is allowed to provide their own buffers for this purpose. This way they can re-use a buffer not needed when performing FOTA if they are limited on RAM.